Updated Cortex architecture doc and documented the experimental blocks storage (#1945)
cortexproject/cortex: docs/architecture.md

Cortex requires that each HTTP request bear a header specifying a tenant ID for the request. Request authentication and authorization are handled by an external reverse proxy.

Incoming samples (writes from Prometheus) are handled by the [distributor](#distributor) while incoming reads (PromQL queries) are handled by the [querier](#querier) or optionally by the [query frontend](#query-frontend).

## Storage

Cortex currently supports two storage engines to store and query the time series:

- Chunks (default, stable)
- Blocks (experimental)

The two engines mostly share the same Cortex architecture with few differences outlined in the rest of the document.

### Chunks storage (default)

> The chunks storage stores each single time series into a separate object called chunk. Each chunk contains the timestamp-value pairs for a given period (defaults to 12 hours) of the series. Series chunks are then indexed by time range period and metric labels, in order to provide a fast lookup across millions - or even billions - of chunks.

Reviewer: ```suggestion
The Chunks storage stores each single time series into a separate object called _Chunk_. Each Chunk contains the timestamp-value pairs for a given period (defaults to 12 hours) of the series. Chunks are then indexed by time range and labels, in order to provide a fast lookup across many (over millions) chunks.
```

Reviewer: Sorry if this is a bit invasive, in this paragraph you're describing a Chunk and its purpose. Seems like in this case a Chunk becomes the subject of the paragraph.

[Prometheus describes](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/chunks.md#chunk) a Chunk as a proper noun. The only case where it doesn't is as it refers to chunk files. I know these are not TSDB Chunks but I'm using that documentation as an example.

Author: Good point, thanks!

For this reason, the chunks storage consists of:

* An index for the chunks. This index can be backed by:
    * [Amazon DynamoDB](https://aws.amazon.com/dynamodb)
    * [Google Bigtable](https://cloud.google.com/bigtable)
    * [Apache Cassandra](https://cassandra.apache.org)
* An object store for the chunk data itself, which can be:
    * [Amazon DynamoDB](https://aws.amazon.com/dynamodb)
    * [Google Bigtable](https://cloud.google.com/bigtable)
    * [Apache Cassandra](https://cassandra.apache.org)
    * [Amazon S3](https://aws.amazon.com/s3)
    * [Google Cloud Storage](https://cloud.google.com/storage/)

Internally, the access to the chunks storage relies on a unified interface called "chunks store". Unlike other Cortex components, the chunk store is not a separate service, but rather a library embedded in the services that need to access the long-term storage: [ingester](#ingester), [querier](#querier) and [ruler](#ruler).

> The chunk and index data format is versioned, in order to allow Cortex operators to upgrade the cluster to the latest stable version and take advantage of new features and improvements - which may introduce changes in the storage format - without requiring any downtime or complex procedures to rewrite the stored data.

Reviewer: ```suggestion
The chunk and index format are versioned, this allows Cortex operators to upgrade the cluster to take advantage of new features and improvements. This strategy enables changes in the storage format without requiring any downtime or complex procedures to rewrite the stored data.
```

> A set of storage schemas are used to map the version to use to read and write time series belonging to a specific period of time. The current schema recommendation is the **v10 schema** (v11 is still experimental). For more information about the schema, please check out the "[Schema](guides/running.md#schema)" documentation.

Reviewer: ```suggestion
A set of schemas are used to map the version while reading and writing time series belonging to a specific period of time. 

The current stable schema is the **v10 schema** (v11 is still experimental). For more information about the schema, please check out the "[Schema](guides/running.md#schema)" documentation.
```

Reviewer: How do you feel about this split? Furthermore, I think the first paragraph here needs to be incorporated into the previous one. In the previous paragraph, we talk about the benefits of the schema but then on this paragraph we define how the schema is used if my understanding is correct which feels counterintuitive.

I think we could do something like:

> The chunk and index format are versioned, this allows Cortex operators to upgrade the cluster to take advantage of new features and improvements. This strategy enables changes in the storage format without requiring any downtime or complex procedures to rewrite the stored data. The set of schemas are used to map the version while reading and writing time series belonging to a specific period of time.

> The current stable schema is the **v10 schema** (v11 is still experimental). For more information about the schema, please refer to the [Schema documentation](guides/running.md#schema).

Author: The grouping you propose makes sense to me. I picked your updated sentence and re-grouped it in the doc.

### Blocks storage (experimental)

> The blocks storage is based on [Prometheus TSDB](https://prometheus.io/docs/prometheus/latest/storage/): it stores all the time series for a given tenant and period (defaults to 2 hours) into a single TSDB block. Each block is composed by few files storing the chunks and the block index.

Reviewer: ```suggestion
The blocks storage is based on [Prometheus TSDB](https://prometheus.io/docs/prometheus/latest/storage/): It stores all the time series for a given tenant and period (defaults to 2 hours) into a single TSDB block. Each Block is composed of files storing the Chunks and the Block Index. You can refer to the [Prometheus documentation for a reference of the files used](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/README.md). 
```

Author: Resolving the comment, cause Thor suggested to completely rephrase it (see new version). Please re-open the comment if you have further comments on it.

> The blocks storage is based on [Prometheus TSDB](https://prometheus.io/docs/prometheus/latest/storage/): it stores all the time series for a given tenant and period (defaults to 2 hours) into a single TSDB block. Each block is composed by few files storing the chunks and the block index.

Reviewer 2: > time series for a given tenant and period (defaults to 2 hours) into a single TSDB block

The wording on this reads weird to me, makes it sound like all users are in a single TSDB. Maybe change to something like "it stores each tenant's time series into their own TSDB which write out their metrics to a on-disk block"

> The chunk files contain the timestamp-value pairs for multiple series. The series inside the chunks are then indexed by a per-block index, which indexes metric names and labels to time series in the chunk files.

Reviewer: When we talk about chunks, are we referring to the same chunks previously defined or TSDB chunks? Would be good to clear up that distinction and link to https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/chunks.md if necessary.

Author: The chunks in the blocks storage are TSDB chunks. Conceptually they're not much different from the "chunks storage" given they both contain timestamp-value pairs, but in TSDB a single chunks file contain multiple series (many).

I've changed it to "TSDB chunk files" without repeating "TSDB chunk" in every single place, cause looks a bit overkilled to me.

> The blocks storage doesn't require a dedicated backend for the index, so it only requires an object store for the block files, which can be:

Reviewer: ```suggestion
The Blocks storage doesn't require a dedicated storage backend for the index. The only requirement is an object store for the block files, which can be:
```

* [Amazon S3](https://aws.amazon.com/s3)
* [Google Cloud Storage](https://cloud.google.com/storage/)

> For more information, please check out the "[Blocks storage](operations/blocks-storage.md)" documentation.

Reviewer: ```suggestion
For more information, please check out the [Blocks storage](operations/blocks-storage.md) documentation.
```

## Services

> Cortex has a service-based architecture, in which the overall system is split up into a variety of components that perform specific tasks and run separately and in parallel.

Reviewer: ```suggestion
Cortex has a service-based architecture, in which the overall system is split up into a variety of components that perform a specific task, run separately, and in parallel.
```

Cortex is, for the most part, a shared-nothing system. Each layer of the system can run multiple instances of each component and they don't coordinate or communicate with each other within that layer.

The Cortex services are:

- [Distributor](#distributor)
- [Ingester](#ingester)
- [Querier](#querier)
- [Query frontend](#query-frontend) (optional)
- [Ruler](#ruler) (optional)
- [Alertmanager](#alertmanager) (optional)

### Distributor

> The **distributor** service is responsible for handling incoming series by Prometheus. It's the first stop in the write path for series samples. Once the distributor receives samples from Prometheus, each series is validated for correctness and to ensure that it is within the configured tenant (or global) limits. Valid samples and then split into batches and sent to multiple [ingesters](#ingester) in parallel.

Author: I've tried to address both feedback. What's about now?

> The **distributor** service is responsible for handling incoming series by Prometheus. It's the first stop in the write path for series samples. Once the distributor receives samples from Prometheus, each series is validated for correctness and to ensure that it is within the configured tenant (or global) limits. Valid samples and then split into batches and sent to multiple [ingesters](#ingester) in parallel.

Reviewer: Looks wonderful 👌

Distributors are **stateless** and can be scaled up and down as needed.

#### HA Tracker

The distributor features an **HA Tracker**. When enabled, the distributor deduplicates incoming samples from redundant Prometheus servers. This allows you to have multiple HA replicas of the same Prometheus servers, writing the same series to Cortex and then deduplicate these series in the Cortex distributor.

> The HA Tracker deduplicates incoming samples based on a cluster and replica label. The cluster label uniquely identifies the cluster of redundant Prometheus servers for a given tenant, while the replica label uniquely identify the replica within the Prometheus servers cluster. Incoming samples are considered duplicated - and thus skipped - if received by any replica which is not the current primary within a cluster.

Author: I think dropped is more correct. I change it, thanks.

The HA Tracker requires a key-value (KV) store to coordinate which is the currently elected primary replica from which the distributor accepts the samples. Samples with one or neither of the replica and cluster labels are accepted by default and never deduplicated.

The supported KV stores for the HA tracker are:

* [Consul](https://www.consul.io)
* [Etcd](https://etcd.io)

For more information, please check out the [Config for sending HA pairs data to Cortex](guides/ha-pair-handling.md) documentation.

#### Hashing

> Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which instances of the ingester service should receive a given series.

Author: Due to the replication factor they're multiple instances, so I think `instances` is more appropriate. I change it, then if you have further feedback please comment.

> Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which instances of the ingester service should receive a given series.

Reviewer: I would use instance(s) then, as there is a chance (when is 1) that they only end up in 1.

> Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which instances of the ingester service should receive a given series.

Author: Done

Cortex supports two hashing strategies:

1. Hash the metric name and tenant ID (default)
2. Hash the metric name, labels and tenant ID (recommended, enabled with `-distributor.shard-by-all-labels=true`)

> The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.

Author: I agree on being clear about the reason.

> increases the likelihood of a query needing

It's not that it increases the likelihood, but we literally always query all ingesters. I've cherry picked your change and changed that part.

Feel free to further comment if you have any other feedback.

> The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.

Reviewer: Thank you for clearing that up! Looks good 👌 

The diff still shows the previous paragraph, not sure if this is indented or not but was worth mentioning.

> The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.

Reviewer: one tiny nit: Is that we use the word "spread" followed by "any" which suggests a division by 1. My suggestion is:

> since a metric could be spread across multiple ingesters

If my understanding of how this works is correct.

> The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.

Author: > The diff still shows the previous paragraph, not sure if this is indented or not but was worth mentioning.

Ops, fixed. Thanks!

> since a metric could be spread across multiple ingesters

Way better.

#### The hash ring

> 

Reviewer: In my opinion, this section is one slightly more complex pieces of Cortex. You've done a fantastic job trying to expand on it, but I'm not sure about the way it reads.

If you feel strongly about it we can leave it as is, if you have doubts - I'm happy to take a stab and try to reword it and perhaps add a small diagram to it. What do you think?

Author: I never feel very strong about something, and I'm always open to discussion 😉 That being said, I would appreciate if could express the same concept in a easier way.

About the diagram, sure. I also thought about it, but I didn't (yet) because I'm not sure how to draw a diagram in a way which will be easily editable by other contributors in the future (ie. which tool to use? where should we share the sources?).

Reviewer: I tend to use [mermaid](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbkFbQ2hyaXN0bWFzXSAtLT58R2V0IG1vbmV5fCBCKEdvIHNob3BwaW5nKVxuQiAtLT4gQ3tMZXQgbWUgdGhpbmt9XG5DIC0tPnxPbmV8IERbTGFwdG9wXVxuQyAtLT58VHdvfCBFW2lQaG9uZV1cbkMgLS0-fFRocmVlfCBGW2ZhOmZhLWNhciBDYXJdXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9fQ) for that exact same reasons. The source code is extremely simple which makes it easy to check it into version control and they have a live editor which means anyone can copy/paste then start editing right away.

Let's merge this as is and I'll create a follow-up issue assigned to me to come up with something.

Reviewer: I have created #1949 to record this conversation. Let me know what you think.

Author: I think it's good not blocking this PR with the improved ring description. I will comment the issue.

Write de-amplification is the main source of Cortex's low total cost of ownership (TCO).

### Querier

The **querier** service handles queries using the [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) query language.

Queriers fetch series samples both from the ingesters and long-term storage: the ingesters hold the in-memory series which have not yet been flushed to the long-term storage. Because of the replication factor, it is possible that the querier may receive duplicated samples; to resolve this, for a given time series the querier internally **deduplicates** samples with the same exact timestamp.

Queriers are **stateless** and can be scaled up and down as needed.

### Query frontend

The **query frontend** is an **optional service** providing the same querier's API endpoints and can be used to accelerate the read path. When the query frontend is in place, incoming query requests should be directed to the query frontend instead of the queriers. The querier service will be still required within the cluster, in order to execute the actual queries.

> The query frontend internally performs some query mangling and push execution jobs for the queriers to an internal queue. In this setup, queriers act as workers picking jobs from the query frontend queue, executing the jobs and pushing back the result which is then aggregated by the query frontend before returning it to the client who performed the query request. Queriers need to be configured with the query frontend address - via the `-querier.frontend-address` CLI flag - in order to allow them to connect to the query frontends and pick up jobs to execute.

Reviewer 3: ```suggestion
The query frontend internally performs some query adjustments and holds queries in an internal queue. In this setup, queriers act as workers which pull jobs from the queue, execute them, and return them to the query-frontend for aggregation. Queriers need to be configured with the query frontend address - via the `-querier.frontend-address` CLI flag in order to allow them to connect to the query frontends.
```

Query frontends are **stateless**. However, due to how the internal queue works, it's recommended to run a few query frontend replicas to reap the benefit of fair scheduling. Two replicas should suffice in most cases.

#### Queueing

The query frontend queuing mechanism is used to:

* Ensure that large queries - that could cause an out-of-memory (OOM) error in the querier - will be retried on failure. This allows administrators to under-provision memory for queries, or optimistically run more small queries in parallel, which helps to reduce the TCO.
* Prevent multiple large requests from being convoyed on a single querier by distributing them across all queriers using a first-in/first-out queue (FIFO).
* Prevent a single tenant from denial-of-service-ing (DOSing) other tenants by fairly scheduling queries between tenants.

#### Splitting

The query frontend splits multi-day queries into multiple single-day queries, executing these queries in parallel on downstream queriers and stitching the results back together again. This prevents large (multi-day) queries from causing out of memory issues in a single querier and helps to execute them faster.

#### Caching

> The query frontend supports Memcached as backend to cache query results and reuses them on subsequent queries. If the cached results are incomplete, the query frontend calculates the required subqueries and executes them in parallel on downstream queriers. The query frontend can optionally align queries with their step parameter to improve the cacheability of the query results.

Reviewer 3: I believe it's compatible with any of our caching backends as it uses `cache.Config` [internally](https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/results_cache.go#L27). The list would be `memcached`, `redis`, `memory`.

> The query frontend supports Memcached as backend to cache query results and reuses them on subsequent queries. If the cached results are incomplete, the query frontend calculates the required subqueries and executes them in parallel on downstream queriers. The query frontend can optionally align queries with their step parameter to improve the cacheability of the query results.

Reviewer 3: ```suggestion
The query frontend supports caching query results and reuses them on subsequent queries. If the cached results are incomplete, the query frontend calculates the required subqueries and executes them in parallel on downstream queriers. The query frontend can optionally align queries with their step parameter to improve the cacheability of the query results. The result cache is compatible with any cortex caching backend (currently memcached, redis, and an in-memory cache).
```

I believe it's compatible with any of our caching backends as it uses `cache.Config` [internally](https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/results_cache.go#L27). The list would be `memcached`, `redis`, `memory`.

The current schema recommendation is the **v10 schema** (v11 is still experimental). For more information about the schema, please check out the [Schema](guides/running.md#schema) documentation.

### Blocks storage (experimental)

The blocks storage is based on [Prometheus TSDB](https://prometheus.io/docs/prometheus/latest/storage/): it stores each tenant's time series into their own TSDB which write out their series to a on-disk Block (defaults to 2h block range periods). Each Block is composed by few files storing the chunks and the block index.

The TSDB chunk files contain the timestamp-value pairs for multiple series. The series inside the Chunks are then indexed by a per-block index, which indexes metric names and labels to time series in the chunk files.

The blocks storage doesn't require a dedicated storage backend for the index. The only requirement is an object store for the Block files, which can be:

* [Amazon S3](https://aws.amazon.com/s3)
* [Google Cloud Storage](https://cloud.google.com/storage/)

For more information, please check out the [Blocks storage](operations/blocks-storage.md) documentation.

## Services

> Cortex has a service-based architecture, in which the overall system is split up into a variety of components that perform a specific task, run separately, and in parallel.

Author: Right. What's about now?

#### Hashing

Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which ingester instances should receive a given series.

Cortex supports two hashing strategies:

1. Hash the metric name and tenant ID (default)
2. Hash the metric name, labels and tenant ID (recommended, enabled with `-distributor.shard-by-all-labels=true`)

The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.

The trade-off associated with the latter is that writes are more balanced across ingesters, but requires each query to look across all ingesters, since a metric could be spread across any ingester given different label sets.

> The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying most of the ingesters.

Author: I simplified it, removing "opinions" and just leaving the trade-off mention.

Ingesters are **semi-stateful**.

#### Ingesters failure and data loss

If an ingester process crashes or exits abruptly, all the in-memory series that have not yet been flushed to the long-term storage will be lost. There are two main ways to mitigate this failure mode:

1. Replication
2. Write-ahead log (WAL)

The **replication** is used to hold multiple (typically 3) replicas of each time series in the ingesters. If the Cortex cluster looses an ingester, the in-memory series hold by the lost ingester are also replicated at least to another ingester. In the event of a single ingester failure, no time series samples will be lost while, in the event of multiple ingesters failure, time series may be potentially lost if failure affects all the ingesters holding the replicas of a specific time series.

The **write-ahead log** (WAL) is used to write to a persistent local disk all incoming series samples until they're flushed to the long-term storage. In the event of an ingester failure, a subsequent process restart will replay the WAL and recover the in-memory series samples.

> Contrary to the sole replication and given the persistent local disk data is not lost, in the event of multiple ingesters failure each ingester will recover the in-memory series samples from WAL upon subsequent restart. The replication is still recommended in order to ensure no temporary failures on the read path in the event of a single ingester failure.

Author: It means, "differently then when you setup a cluster only with the replication and without the WAL"
